agent/lib/kontena/workers/log_worker.rb

```diff
@@ -46,6 +47,7 @@ def initialize(autostart = true)
 
     def watch_queue
       every(WATCH_INTERVAL) do
+        flush_buffer if force_flush_buffer?
```

Comment:
Maybe `flush_buffer if force_flush_buffer? && @processing`? Force flushing without `@processing` being `true` could flush things to non-connected rpc pipe :/ Eventually those would timeout and logs go to `/dev/null`. Also in cases with high log volumes and short downtime for server this could potentially put few rpc notifications with lot of log entries (that's since flush_buffer takes full buffer without chunking it) into the pipe that would then spuriously hit the master when connected if they do not timeout.

Reply:
Calling `flush_buffer` from both the actor `watch_queue` => `every` task and the `process_queue` => `defer` thread risks races on the `@buffer`. There should only be one thread handling it.

I think I'd consider getting rid of the `defer` thread, and try replacing the blocking `@queue.pop` with non-blocking pops from a timer.

Reply:
Removed `defer` and replaced `Queue` with `Array` so that it's possible to `shift` items from queue in batches. Array should be fine here because each `ContainerLogWorker` ensures ordering for a container. This simplified logic quite nicely.